### PR TITLE
feat: add tracing framework

### PR DESCRIPTION
**Pull request by ianbotsf**
## Issue \#

Closes [smithy-kotlin#677](https://github.com/awslabs/smithy-kotlin/issues/677)

## Description of changes

These are the **aws-sdk-kotlin** changes to adopt the new tracing framework introduced in [smithy-kotlin#737](https://github.com/awslabs/smithy-kotlin/pull/737).

**Note**: This is the final, mergeable version of #731, including changes from the rejected approaches in #680 (context receivers) and #708 (ExecutionContext).

**Companion PR**: [smithy-kotlin#737](https://github.com/awslabs/smithy-kotlin/pull/737)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/StsAssumeRoleCredentialsProvider.kt [hunk 1]
```diff
@@ -64,6 +66,7 @@ public class StsAssumeRoleCredentialsProvider(
             region = provider.region ?: GLOBAL_STS_PARTITION_ENDPOINT
             credentialsProvider = provider.credentialsProvider
             httpClientEngine = provider.httpClientEngine
+            tracer = traceSpan.asNestedTracer("STS-")
```
Comment (aajtodd) [inline aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/StsAssumeRoleCredentialsProvider.kt:69]: question/nit: Perhaps the trailing `-` separator should be handled internally?

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/imds/TokenMiddleware.kt [hunk 2]
```diff
@@ -64,7 +65,7 @@ internal class TokenMiddleware(
             }
         }
 
-        val call = httpClient.call(tokenReq)
+        val call = httpClient.call(SdkHttpRequest(tokenReq))
```
Comment (aajtodd) [inline aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/imds/TokenMiddleware.kt:68]: question: Where is this change stemming from?

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/CredentialsProviderChain.kt [hunk 3]
```diff
@@ -23,20 +25,21 @@ import aws.smithy.kotlin.runtime.logging.Logger
...
     init {
         require(providers.isNotEmpty()) { "at least one provider must be in the chain" }
     }
 
     override fun toString(): String =
         (listOf(this) + providers).map { it::class.simpleName }.joinToString(" -> ")
 
-    override suspend fun getCredentials(): Credentials {
-        val chainException = lazy { CredentialsProviderException("No credentials could be loaded from the chain: $this") }
+    override suspend fun getCredentials(): Credentials = coroutineContext.withChildTraceSpan("Credentials chain") {
```
Comment (aajtodd) [inline aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/CredentialsProviderChain.kt:35]: comment: This isn't listed in your span taxonomy FYI (I think it's fine here just mentioning it)

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/StsAssumeRoleCredentialsProvider.kt (hunk 1 above)
Comment (ianbotsf) [reply aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/StsAssumeRoleCredentialsProvider.kt:69]: Good idea.

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/imds/TokenMiddleware.kt (hunk 2 above)
Comment (ianbotsf) [reply aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/config/imds/TokenMiddleware.kt:68]: This was accidentally left over from a now-removed refactor. I'll revert it.

File: aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/CredentialsProviderChain.kt (hunk 3 above)
Comment (ianbotsf) [reply aws-runtime/aws-config/common/src/aws/sdk/kotlin/runtime/auth/credentials/CredentialsProviderChain.kt:35]: Yep you're right, I forgot I left this here. I'll update the taxonomy definition.

I'll also be posting the taxonomy in a new section of **tracing.md** in [the **smithy-kotlin** PR](https://github.com/awslabs/smithy-kotlin/pull/737). I'll make sure to include this there as well.

**Review by lauzadis — APPROVED** [review]

**Review by lucix-aws — APPROVED** [review]

**Pull request merged**